util/etcd/etcd.go

```diff
@@ -0,0 +1,382 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"net"
+	"time"
+
+	"github.com/go-logr/logr"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const etcdTimeout = 2 * time.Second
+
+// exponential backoff for etcd operations, values taken from kubeadm
+var etcdBackoff = wait.Backoff{
+	Steps:    9,
+	Duration: 50 * time.Millisecond,
+	Factor:   2.0,
+	Jitter:   0.1,
+}
+
+type grpcDial = func(ctx context.Context, addr string) (net.Conn, error)
+
+// Client provides connection parameters for an etcd cluster
+type Client struct {
+	dialerContext grpcDial
+	tlsConfig     *tls.Config
+	log           logr.Logger
+	timeout       time.Duration
+	backoffParams wait.Backoff
+	endpoint      string
+}
+
+// MemberAlarm represents an alarm type association with a cluster member
+type MemberAlarm struct {
+	// MemberID is the ID of the member associated with the raised alarm
+	MemberID uint64
+
+	// Type is the type of alarm which has been raised.
+	Type AlarmType
+}
+
+type AlarmType int32
+
+const (
+	// AlarmOK denotes that the cluster member is OK
+	AlarmOk AlarmType = iota
+	// AlarmNoSpace denotes that the cluster member has run out of disk space
+	AlarmNoSpace AlarmType = iota
```

Comment:
Honestly i didn't even realize you could do `= iota` like this. If you wanted, you could remove the `= iota` on all but the first like this:

```
const (
	 AlarmOk AlarmType = iota
	AlarmNoSpace AlarmType
	AlarmCorrupt AlarmType
```

(last example here https://golang.org/ref/spec#Constant_declarations) 

But this is also fine if you prefer the `= iota` all the way down. Both are fine.

Reply:
Aren't all the constants equivalent and indistinguishable if you use `iota` for all of them?

Reply:
Apparently not, I took the first example from https://golang.org/ref/spec#Iota

Reply:
TIL!

Reply:
Me too! Something _iota_ have known.